expression/builtin_time.go

```diff
@@ -3280,1581 +3388,616 @@ func (du *baseDateArithmetical) vecGetIntervalFromReal(b *baseBuiltinFunc, input
+	// However, MySQL has a weird behavior for date_add(string, ...), whose result depends on the content of the first argument.
+	// E.g., date_add('2000-01-02 00:00:00', interval 1 day) evaluates to '2021-01-03 00:00:00' (which is normal),
+	// whereas date_add('2000-01-02', interval 1 day) evaluates to '2000-01-03' instead of '2021-01-03 00:00:00'.
+	// This requires a customized parsing of the content of the first argument, by recognizing if it is a pure date format or contains HMS part.
+	// So implicit casts are not viable here.
+	if dateEvalTp == types.ETTimestamp {
 		dateEvalTp = types.ETDatetime
+	} else if dateEvalTp == types.ETJson {
+		dateEvalTp = types.ETString
 	}
 
 	intervalEvalTp := args[1].GetType().EvalType()
-	if intervalEvalTp != types.ETString && intervalEvalTp != types.ETDecimal && intervalEvalTp != types.ETReal {
+	if intervalEvalTp == types.ETJson {
+		intervalEvalTp = types.ETString
+	} else if intervalEvalTp != types.ETString && intervalEvalTp != types.ETDecimal && intervalEvalTp != types.ETReal {
 		intervalEvalTp = types.ETInt
 	}
 
-	argTps := []types.EvalType{dateEvalTp, intervalEvalTp, types.ETString}
-	var bf baseBuiltinFunc
 	unit, _, err := args[2].EvalString(ctx, chunk.Row{})
 	if err != nil {
 		return nil, err
 	}
-	if dateEvalTp == types.ETDuration {
 
-		internalFsp := 0
-		switch unit {
-		// If the unit has micro second, then the fsp must be the MaxFsp.
-		case "MICROSECOND", "SECOND_MICROSECOND", "MINUTE_MICROSECOND", "HOUR_MICROSECOND", "DAY_MICROSECOND":
-			internalFsp = types.MaxFsp
-		// If the unit is second, the fsp is related with the arg[1]'s.
-		case "SECOND":
-			internalFsp = types.MaxFsp
-			if intervalEvalTp != types.ETString {
-				internalFsp = mathutil.Min(args[1].GetType().GetDecimal(), types.MaxFsp)
-			}
-			// Otherwise, the fsp should be 0.
-		}
-		bf, err = newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETDuration, argTps...)
-		if err != nil {
-			return nil, err
-		}
-		arg0Dec, err := getExpressionFsp(ctx, args[0])
-		if err != nil {
-			return nil, err
-		}
-		bf.setDecimalAndFlenForTime(mathutil.Max(arg0Dec, internalFsp))
-	} else if dateEvalTp == types.ETString {
-		bf, err = newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, argTps...)
-		if err != nil {
-			return nil, err
-		}
-		bf.tp.SetFlen(mysql.MaxDatetimeFullWidth)
-	} else {
-		bf, err = newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETDatetime, argTps...)
-		if err != nil {
-			return nil, err
-		}
-		if dateEvalTp == types.ETDatetime && args[0].GetType().GetType() == mysql.TypeTimestamp {
-			tp := types.NewFieldType(mysql.TypeDatetime)
-			tp.SetDecimal(args[0].GetType().GetDecimal())
-			tp.SetFlen(mysql.MaxDatetimeWidthNoFsp)
-			if tp.GetDecimal() > 0 {
-				tp.SetFlen(tp.GetFlen() + 1 + tp.GetDecimal())
-			}
-			args[0] = BuildCastFunction(ctx, args[0], tp)
+	resultTp := mysql.TypeVarString
+	resultEvalTp := types.ETString
+	if args[0].GetType().GetType() == mysql.TypeDate {
```

Comment:
```suggestion
	if dateEvalTp == mysql.TypeDate {
```

Reply:
I'm afraid this is not valid. There are two type systems in TiDB: "eval type" and "MySQL type". They are for different purpose and not supposed to be mix-used.

Eval type is more like "how data is represented during evaluation", for example, all integer columns, enum columns, and set columns have `ETInt`, and both date and datetime columns have `ETDatetime`.

The examples above reveal exactly the tricks in type inference:
1. Sometimes it is handy to just infer types from eval type - "if this argument has ETInt, no matter it is smallint, bigint, or even enum, the overload is resolved to be this one".
2. Sometimes a function may have a fine-grained type inference rule - "if this argument is date type, the return type is date. If the first argument is datetime type, the return type is datetime". Now the eval type is insufficient. We need to fall back to consulting the mysql type, which is the type specified in the create table statement. And this is what this line does.

Reply:
Ah, thank you for the explanation and sorry for missing it :(